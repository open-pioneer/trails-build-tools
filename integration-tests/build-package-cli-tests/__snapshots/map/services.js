export { MapRegistryImpl } from './MapRegistryImpl.js';
//# sourceMappingURL=services.js.map
