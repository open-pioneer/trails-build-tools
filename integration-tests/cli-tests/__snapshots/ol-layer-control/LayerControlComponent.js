import { jsx, jsxs } from 'react/jsx-runtime';
import { ScaleFade, Checkbox, Slider, SliderTrack, SliderFilledTrack, Tooltip, SliderThumb } from '@open-pioneer/chakra-integration';
import { useMap } from '@open-pioneer/experimental-ol-map';
import { unByKey } from 'ol/Observable';
import { useIntl } from './_virtual/_virtual-pioneer-module_react-hooks.js';
import { useMemo, useState, useEffect } from 'react';
import { useTimeout } from 'react-use';

function LayerControlComponent(config) {
  const intl = useIntl();
  const { loading, error, map } = useMap(config.mapId);
  const layers = useMemo(() => map?.getAllLayers().reverse() ?? [], [map]);
  const [hasTimeoutElapsed] = useTimeout(100);
  const fadeIn = !loading || hasTimeoutElapsed() || false;
  return /* @__PURE__ */ jsx(ScaleFade, { in: fadeIn, children: loading ? /* @__PURE__ */ jsx("div", { children: intl.formatMessage({ id: "loading" }) }) : error ? /* @__PURE__ */ jsxs("div", { children: [
    intl.formatMessage({ id: "error" }),
    " ",
    error.message
  ] }) : /* @__PURE__ */ jsx("div", { children: layers.map((layer, i) => /* @__PURE__ */ jsxs("div", { className: "layer-entry", children: [
    /* @__PURE__ */ jsx(
      LayerVisibilityTogglerComponent,
      {
        layer
      }
    ),
    config.showOpacitySlider && /* @__PURE__ */ jsx(
      LayerOpacitySliderComponent,
      {
        layer
      }
    )
  ] }, i)) }) });
}
function LayerVisibilityTogglerComponent(props) {
  const intl = useIntl();
  const [visibility, setVisibility] = useState(props.layer.getVisible());
  const title = props.layer.getProperties().title ?? intl.formatMessage({ id: "undefined-layer-title" });
  const changeVisibility = () => {
    setVisibility(!visibility);
    props.layer.setVisible(!visibility);
  };
  return /* @__PURE__ */ jsx(
    Checkbox,
    {
      className: "layer-select",
      size: "lg",
      isChecked: visibility,
      onChange: changeVisibility,
      children: title
    }
  );
}
function LayerOpacitySliderComponent(props) {
  const [sliderValue, setSliderValue] = useState(props.layer.getOpacity() * 100);
  const [showTooltip, setShowTooltip] = useState(false);
  useEffect(() => {
    const opacityChangeListener = props.layer.on("change:opacity", () => {
      const opacity = props.layer.getOpacity() * 100;
      if (opacity !== sliderValue) {
        setSliderValue(Math.round(opacity));
      }
    });
    return () => unByKey(opacityChangeListener);
  }, []);
  return /* @__PURE__ */ jsxs(
    Slider,
    {
      id: "slider",
      value: sliderValue,
      min: 0,
      max: 100,
      colorScheme: "teal",
      onChange: (v) => props.layer.setOpacity(v / 100),
      onMouseEnter: () => setShowTooltip(true),
      onMouseLeave: () => setShowTooltip(false),
      children: [
        /* @__PURE__ */ jsx(SliderTrack, { children: /* @__PURE__ */ jsx(SliderFilledTrack, {}) }),
        /* @__PURE__ */ jsx(
          Tooltip,
          {
            hasArrow: true,
            bg: "teal.500",
            color: "white",
            placement: "top",
            isOpen: showTooltip,
            label: `${sliderValue}%`,
            children: /* @__PURE__ */ jsx(SliderThumb, { bg: "teal.500" })
          }
        )
      ]
    }
  );
}

export { LayerControlComponent };
//# sourceMappingURL=LayerControlComponent.js.map
