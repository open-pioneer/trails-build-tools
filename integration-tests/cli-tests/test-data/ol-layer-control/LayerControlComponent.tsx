// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
import {
    Checkbox,
    ScaleFade,
    Slider,
    SliderFilledTrack,
    SliderThumb,
    SliderTrack,
    Tooltip
} from "@open-pioneer/chakra-integration";
import { OlComponentProps, useMap } from "@open-pioneer/experimental-ol-map";
import Layer from "ol/layer/Layer";
import { unByKey } from "ol/Observable";
import { useIntl } from "open-pioneer:react-hooks";
import { useEffect, useMemo, useState } from "react";
import { useTimeout } from "react-use";

export interface LayerControlProps extends OlComponentProps {
    /**
     * Sets visibility of opacity slider
     */
    showOpacitySlider?: boolean;
}

export function LayerControlComponent(config: LayerControlProps) {
    const intl = useIntl();
    const { loading, error, map } = useMap(config.mapId);
    const layers = useMemo(() => map?.getAllLayers().reverse() ?? [], [map]);

    // Small timeout before "Loading..." to optimize for the common case where the map
    // is available almost immediately. This prevents some flickering in the UI.
    const [hasTimeoutElapsed] = useTimeout(100);
    const fadeIn = !loading || hasTimeoutElapsed() || false;
    return (
        <ScaleFade in={fadeIn}>
            {loading ? (
                <div>{intl.formatMessage({ id: "loading" })}</div>
            ) : error ? (
                <div>
                    {intl.formatMessage({ id: "error" })} {error.message}
                </div>
            ) : (
                <div>
                    {layers.map((layer, i) => (
                        <div key={i} className="layer-entry">
                            <LayerVisibilityTogglerComponent
                                layer={layer}
                            ></LayerVisibilityTogglerComponent>
                            {config.showOpacitySlider && (
                                <LayerOpacitySliderComponent
                                    layer={layer}
                                ></LayerOpacitySliderComponent>
                            )}
                        </div>
                    ))}
                </div>
            )}
        </ScaleFade>
    );
}

function LayerVisibilityTogglerComponent(props: { layer: Layer }) {
    const intl = useIntl();
    const [visibility, setVisibility] = useState<boolean>(props.layer.getVisible());
    const title =
        props.layer.getProperties().title ?? intl.formatMessage({ id: "undefined-layer-title" });

    const changeVisibility = () => {
        setVisibility(!visibility);
        props.layer.setVisible(!visibility);
    };

    return (
        <Checkbox
            className="layer-select"
            size="lg"
            isChecked={visibility}
            onChange={changeVisibility}
        >
            {title}
        </Checkbox>
    );
}

function LayerOpacitySliderComponent(props: { layer: Layer }) {
    const [sliderValue, setSliderValue] = useState(props.layer.getOpacity() * 100);
    const [showTooltip, setShowTooltip] = useState(false);

    useEffect(() => {
        const opacityChangeListener = props.layer.on("change:opacity", () => {
            const opacity = props.layer.getOpacity() * 100;
            if (opacity !== sliderValue) {
                setSliderValue(Math.round(opacity));
            }
        });
        return () => unByKey(opacityChangeListener);
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);

    return (
        <Slider
            id="slider"
            value={sliderValue}
            min={0}
            max={100}
            colorScheme="teal"
            onChange={(v) => props.layer.setOpacity(v / 100)}
            onMouseEnter={() => setShowTooltip(true)}
            onMouseLeave={() => setShowTooltip(false)}
        >
            <SliderTrack>
                <SliderFilledTrack />
            </SliderTrack>
            <Tooltip
                hasArrow
                bg="teal.500"
                color="white"
                placement="top"
                isOpen={showTooltip}
                label={`${sliderValue}%`}
            >
                <SliderThumb bg="teal.500"></SliderThumb>
            </Tooltip>
        </Slider>
    );
}
