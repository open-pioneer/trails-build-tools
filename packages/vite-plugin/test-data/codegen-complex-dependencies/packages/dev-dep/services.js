export class Service {
    test() {
        console.log("from dev dep");
    }
}
