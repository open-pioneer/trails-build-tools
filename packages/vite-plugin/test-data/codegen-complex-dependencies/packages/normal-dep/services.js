export class Service {
    test() {
        console.log("from normal dep");
    }
}
