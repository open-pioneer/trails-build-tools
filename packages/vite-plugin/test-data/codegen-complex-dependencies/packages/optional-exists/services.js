export class Service {
    test() {
        console.log("from optional dep");
    }
}
