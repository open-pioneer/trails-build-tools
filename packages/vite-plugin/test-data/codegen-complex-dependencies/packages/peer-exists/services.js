export class Service {
    test() {
        console.log("from peer dep");
    }
}
