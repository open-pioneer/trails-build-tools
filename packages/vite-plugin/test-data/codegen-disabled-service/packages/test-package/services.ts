export class A {
    test() {
        console.log("from A");
    }
}

export class B {
    test() {
        console.log("from B");
    }
}
