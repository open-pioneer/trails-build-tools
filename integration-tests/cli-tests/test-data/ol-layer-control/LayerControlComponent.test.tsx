// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
/**
 * @vitest-environment jsdom
 */
import { OlMapConfigurationProvider } from "@open-pioneer/experimental-ol-map";
import { OlMapRegistry } from "@open-pioneer/experimental-ol-map/services";
import { Service, ServiceOptions } from "@open-pioneer/runtime";
import {
    PackageContextProvider,
    PackageContextProviderProps
} from "@open-pioneer/test-utils/react";
import { createService } from "@open-pioneer/test-utils/services";
import { act, render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { MapOptions } from "ol/Map";
import TileLayer from "ol/layer/Tile";
import Stamen from "ol/source/Stamen";
import { expect, it } from "vitest";

import { LayerControlComponent } from "./LayerControlComponent";

// used to avoid a "ResizeObserver is not defined" error
global.ResizeObserver = require("resize-observer-polyfill");

class MapConfigProvider implements OlMapConfigurationProvider {
    mapId = "default";
    mapOptions: MapOptions = {};

    constructor(options: ServiceOptions) {
        if (options.properties.mapOptions) {
            this.mapOptions = options.properties.mapOptions as MapOptions;
        }
        if (options.properties.mapId) {
            this.mapId = options.properties.mapId as string;
        }
    }

    getMapOptions(): Promise<MapOptions> {
        return Promise.resolve(this.mapOptions);
    }
}

async function createOlMapRegistry(mapId: string, mapOptions: MapOptions) {
    const mapConfigProvider = await createService(MapConfigProvider, {
        properties: {
            mapOptions: mapOptions,
            mapId
        }
    });
    return await createService(OlMapRegistry, {
        references: {
            providers: [mapConfigProvider]
        }
    });
}

function createPackageContextProviderProps(
    service: Service<OlMapRegistry>
): PackageContextProviderProps {
    return {
        services: {
            "ol-map.MapRegistry": service
        }
    };
}

it("should successfully create a layer control component", async () => {
    const mapId = "test";
    const mapOptions = {
        layers: [
            new TileLayer({
                source: new Stamen({ layer: "watercolor" }),
                properties: { title: "Watercolor" },
                visible: false
            })
        ]
    } as MapOptions;
    const service = await createOlMapRegistry(mapId, mapOptions);

    const renderResult = render(
        <PackageContextProvider {...createPackageContextProviderProps(service)}>
            <div data-testid="base">
                <LayerControlComponent mapId="test" />
            </div>
        </PackageContextProvider>
    );

    // Assert layer control is mounted
    const div = await screen.findByTestId("base");
    expect(div).toMatchSnapshot();

    // Check if two layer controls for the configured layers are created
    const layerElems = renderResult.container.querySelectorAll(".layer-entry");
    expect(layerElems.length).toBe(1);
    expect(layerElems[0]).toBeInstanceOf(HTMLDivElement);
});

it("layer control should have checkbox to toggle layer visibility", async () => {
    const mapId = "test";
    const mapOptions = {
        layers: [
            new TileLayer({
                source: new Stamen({ layer: "watercolor" }),
                properties: { title: "Watercolor" },
                visible: false
            })
        ]
    } as MapOptions;
    const service = await createOlMapRegistry(mapId, mapOptions);
    const user = userEvent.setup();
    render(
        <PackageContextProvider {...createPackageContextProviderProps(service)}>
            <LayerControlComponent mapId={mapId} showOpacitySlider={true} />
        </PackageContextProvider>
    );

    // pre check visibility of the layer
    const map = await service.getMap(mapId);
    const layers = map.getAllLayers();
    expect(layers.length).toBe(1);
    const firstLayer = layers[0]!;
    expect(firstLayer.getVisible()).toBe(false);

    // adjust visibility by control
    const checkbox = await screen.findByRole("checkbox");
    await user.click(checkbox);
    expect(firstLayer.getVisible()).toBe(true);
});

it("layer control should have usable opacity slider", async () => {
    const mapId = "test";
    const mapOptions = {
        layers: [
            new TileLayer({
                source: new Stamen({ layer: "watercolor" }),
                properties: { title: "Watercolor" },
                visible: false
            })
        ]
    } as MapOptions;
    const service = await createOlMapRegistry(mapId, mapOptions);
    const user = userEvent.setup();
    render(
        <PackageContextProvider {...createPackageContextProviderProps(service)}>
            <LayerControlComponent mapId={mapId} showOpacitySlider={true} />
        </PackageContextProvider>
    );

    // pre check opacity in layer
    const map = await service.getMap(mapId);
    const layers = map.getAllLayers();
    expect(layers.length).toBe(1);
    const firstLayer = layers[0]!;
    expect(firstLayer.getOpacity()).toBe(1);

    // adjust opacity by control
    const slider = await screen.findByRole("slider");
    await act(async () => {
        slider.focus();
        await user.keyboard("[ArrowLeft]");
        expect(firstLayer.getOpacity()).toBeCloseTo(0.99);
    });
});
