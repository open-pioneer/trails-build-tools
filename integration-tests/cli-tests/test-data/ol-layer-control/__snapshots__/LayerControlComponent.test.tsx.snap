// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should successfully create a layer control component 1`] = `
<div
  data-testid="base"
>
  <div
    class="chakra-offset-slide"
    style="opacity: 0; transform: scale(0.95) translateZ(0);"
  >
    <div>
      <div
        class="layer-entry"
      >
        <label
          class="chakra-checkbox layer-select css-1577qb8"
          data-theme="light"
        >
          <input
            class="chakra-checkbox__input"
            style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
            type="checkbox"
            value=""
          />
          <span
            aria-hidden="true"
            class="chakra-checkbox__control css-edb818"
            data-theme="light"
          />
          <span
            class="chakra-checkbox__label css-r90m2q"
            data-theme="light"
          >
            Watercolor
          </span>
        </label>
      </div>
    </div>
  </div>
</div>
`;
